### PR TITLE
v1.8 backports 2020-12-17

### PR DESCRIPTION
**Pull request by jibi**
* #14313 -- pkg/k8s: fix k8s_event_lag_seconds for negative time (@aanm)
 * #14300 -- ipcache: Use controller.Manager on IPIdentityCache for bpf-garbage-collection (@dctrwatson)
 * #14286 -- test: Fix flake on policy verdict count check (@pchaigno)
 * ~#14299 -- Fix rare crash on startup when kubernetes initialization occurs before IP address configuration (@joestringer)~
 * #14375 -- pkg/k8s: fix nil pointer for invalid CCNP (@aanm)
 * #14381 -- ipsec: Fix xfrm output mask (@pchaigno)
 * #14412 -- test: Test policy enforcement through tunnels (@pchaigno)
 * #13465 -- test: Disable host firewall by default when running tests locally (@pchaigno)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 14313 14300 14286 14375 14381 14412 13465; do contrib/backporting/set-labels.py $pr done 1.8; done
```

#### Conflicts tl;dr

* #14286 and #14412 had some trivial conflicts, but please double check them
* #14299: in v1.8, the call to `d.nodeDiscovery.StartDiscovery(nodeTypes.GetName())` is happening after the [annotation of the k8s node logic](https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon.go#L633-L655) rather than before. I don't have enough context to tell if this is OK or not
* #14381: I had to rerun `go mod tidy` and `go mod vendor` as the upstream commit was adding a new dependency

### CODE REVIEW

**Comment by jibi** [comment]
test-backport-1.8

**Comment by jibi** [comment]
test-missed-k8s

**Comment by jibi** [comment]
test-runtime

**Comment by jibi** [comment]
@aanm I think unit tests are failing due to 050fd6799708efb3d3a9a9b27929fb4a921a5b22, does that commit depend on any other change that should also be backported?

File: pkg/k8s/factory_functions_test.go [hunk 1]
```diff
@@ -1048,6 +1048,15 @@ func (s *K8sSuite) Test_ConvertToCNP(c *C) {
...
 		},
+		{
+			name: "A CCNP where it doesn't contain neither a spec nor specs",
+			args: args{
+				obj: &v2.CiliumClusterwideNetworkPolicy{},
+			},
+			want: &types.SlimCNP{
+				CiliumNetworkPolicy: &v2.CiliumNetworkPolicy{},
+			},
+		},
```
Comment (aanm) [inline pkg/k8s/factory_functions_test.go]: You can either drop this change or you also import this change https://github.com/cilium/cilium/commit/c813a15041dd8996a686a218ffae61d74b4ef137#diff-e3ab607a4306e41dfa6064994ce4ca7d1bde3a7adf750133bdd018d4b80c8e55R1106

**Review by aanm — CHANGES_REQUESTED** [review]

**Review by joestringer — APPROVED** [review]
#14299 LGTM for the fix. `StartDiscovery()` is what triggers the `<-d.nodeDiscovery.Registered` channel close which ensures that any regeneration of the datapath can take into account the known node configuration. Making `StartDiscovery()` go later is OK, it just means that generating the datapath will start a little later. As long as it isn't going too early, we shouldn't reintroduce the bug.

I agree the location of StartDiscovery() seems a little fishy since line 567-569 at least pull the health IP from nodeDiscovery before `StartDiscovery()` is called. In the latest code, commit 91ffa1cca9b55649fb573f37a8d79fac7e422d2c actually moved this around without calling it out as a bugfix so the consequences aren't immediately apparent.

I checked all of the fields from line 566-571 and the health IPs are the only ones that aren't guaranteed to be initialized by the point when the k8s annotation is created. Maybe in some scenarios this could cause health IPs to be propagated wrong in the cluster. In that case, some nodes might report connectivity issues to other nodes because they have the wrong IP. Unless we hear user reports of an issue on v1.8 relating to this, I think it's fine to leave as-is.

I also looked at #14300 , LGTM.

**Review by pchaigno — APPROVED** [review]
My PRs look good :+1: 

I've also validated the IPSec fix manually by running our `DatapathConfig Encapsulation` tests with kube-proxy enabled.

**Comment by joestringer** [comment]
`K8sDatapathConfig Etcd Check connectivity` test failed across all jobs, it looks like one of the backports is introducing a regression that is influencing Cilium's ability to start up and serve the API when running against etcd on this branch. #14299 seems like the most likely culprit at a glance.

**Comment by joestringer** [comment]
test-backport-1.8

**Comment by joestringer** [comment]
I dropped bb9f0252d2c8 ("loader: Wait for node configuration to generate datapath") and re-kicked CI to see if that resolves the conflict. If so, we can look a little closer at the backport specifically for that PR.

**Comment by joestringer** [comment]
Unit test failure seems possibly related to the changes:
```
FAIL: factory_functions_test.go:1033: K8sSuite.Test_ConvertToCNP
factory_functions_test.go:1098:
    c.Assert(got, checker.DeepEquals, tt.want, Commentf("Test Name: %s", tt.name))
... obtained *v2.CiliumClusterwideNetworkPolicy = &v2.CiliumClusterwideNetworkPolicy{CiliumNetworkPolicy:(*v2.CiliumNetworkPolicy)(nil), Status:v2.CiliumNetworkPolicyStatus{Nodes:map[string]v2.CiliumNetworkPolicyNodeStatus(nil), DerivativePolicies:map[string]v2.CiliumNetworkPolicyNodeStatus(nil)}} (%!q(PANIC=String method: runtime error: invalid memory address or nil pointer dereference))
... expected *types.SlimCNP = &types.SlimCNP{CiliumNetworkPolicy:(*v2.CiliumNetworkPolicy)(0xc0002dcdc0)} ("TypeMeta: &TypeMeta{Kind:,APIVersion:,}, ObjectMeta: &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[]OwnerReference{},Finalizers:[],ClusterName:,ManagedFields:[]ManagedFieldsEntry{},}, Status: {map[] map[]}")
... Test Name: A CCNP where it doesn't contain neither a spec nor specs
... Unified diff:
--- obtained
+++ expected
@@ -1 +1,3 @@
-&v2.CiliumClusterwideNetworkPolicy{}
+&types.SlimCNP{
+    CiliumNetworkPolicy: &v2.CiliumNetworkPolicy{},
+}
```

**Comment by jibi** [comment]
> I dropped [bb9f025](https://github.com/cilium/cilium/commit/bb9f0252d2c8f4a83e700d56244dc6abcb415ba7) ("loader: Wait for node configuration to generate datapath") and re-kicked CI to see if that resolves the CI breakage. If so, we can look a little closer at the backport specifically for that PR.

yeah, after bisecting the branch I came to the same conclusion. I updated the PR to drop https://github.com/cilium/cilium/commit/bb9f0252d2c8f4a83e700d56244dc6abcb415ba7 and also to include the changes that André suggested to sort unit tests (I updated https://github.com/cilium/cilium/pull/14441/commits/bddf12e5eaf4ec96a95067e313abbf146cf854ed commit message with a note for that)

**Comment by jibi** [comment]
test-backport-1.8

**Comment by jibi** [comment]
test-missed-k8s

**Review by aanm — APPROVED** [review]
My changes LGTM. Thanks

**Comment by jibi** [comment]
`K8sDatapathConfig AutoDirectNodeRoutes Check connectivity with automatic direct nodes routes` is failing. I'm not sure I can pinpoint that failure to a specific commit so I'll try to reproduce this locally and bisect again the branch

**Comment by jibi** [comment]
I added back https://github.com/cilium/cilium/pull/14412 as without that PR there were even more tests failing

**Comment by jibi** [comment]
test-backport-1.8

**Comment by jibi** [comment]
(rebased on top of 1.8 to fix a conflict)

**Comment by jibi** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-missed-k8s

**Comment by jibi** [comment]
test-missed-k8s

(not sure if that is a flake or a genuine failure)

**Comment by jibi** [comment]
test-missed-k8s

**Comment by jibi** [comment]
After bisecting the branch locally it looks like the `K8sDatapathConfig_AutoDirectNodeRoutes_Check_connectivity_with_automatic_direct_nodes_routes` test started failing after introducing 27a60b855cdc944d1d1f7219bdc09eb343f00e90 (/cc @pchaigno which may have more context on that change)

**Comment by jibi** [comment]
test-bakport-1.8

**Comment by jibi** [comment]
test-backport-1.8

**Comment by jibi** [comment]
test-missed-k8s

**Review by pchaigno — APPROVED** [review]
:+1: for the backport that was added last.

For context: we discovered with @jibi that the host firewall was enabled by default in v1.8's k8s-all CI pipeline. That was causing tests to fail due to known issue #12205. Backporting #13465 disables the host firewall by default in CI.

**Comment by pchaigno** [comment]
Travis CI is currently failing on v1.8; #14531 should fix it. In the meantime, I've manually run the unit tests. Marking as ready-to-merge and merging.

**Comment by pchaigno** [comment]
Labels updated.

**Pull request merged**